### PR TITLE
Update `System.Contract.Call`

### LINKED ISSUE #1646: Update `System.Contract.Call`
Port changes from https://github.com/neo-project/neo/pull/2166, https://github.com/neo-project/neo/pull/2167 .
Effectively, `CallEx` is removed and `Call` is altered.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1646 .

### CODE REVIEW

**Comment by fyrchik** [comment]
There is a problem with RPC: we need to know if invoked method will return any value. Right now we always assume "yes".

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

**Comment by AnnaShaleva** [comment]
> There is a problem with RPC: we need to know if invoked method will return any value. Right now we always assume "yes".

We may add one more parameter to the RPC call, can we create an issue for that in C# repo?

**Comment by roman-khimov** [comment]
> There is a problem with RPC: we need to know if invoked method will return any value. Right now we always assume "yes".

Do you mean `CreateFunctionInvocationScript`? It's not hard to fix it, other than that I probably don't see any problematic cases, when we call some specific methods from client we do know what to expect from them and server-side we just return the stack from test invocations as is whether it has something or not.

But see neo-project/neo-node#710 also.

File: pkg/core/interop/contract/call.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 
 	"github.com/nspcc-dev/neo-go/pkg/core/interop"
 	"github.com/nspcc-dev/neo-go/pkg/core/state"
-	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+	flags2 "github.com/nspcc-dev/neo-go/pkg/smartcontract/flags"
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/interop/contract/call.go]: Fixed.

**Comment by fyrchik** [comment]
RPC problem is no longer an issue, after neo-project/neo#2211 .

File: pkg/core/interops.go [hunk 2]
```diff
@@ -51,9 +51,7 @@ var systemInterops = []interop.Function{
 	{Name: interopnames.SystemBlockchainGetTransactionHeight, Func: bcGetTransactionHeight, Price: 1 << 15,
 		RequiredFlags: flags.ReadStates, ParamCount: 1},
 	{Name: interopnames.SystemContractCall, Func: contract.Call, Price: 1 << 15,
-		RequiredFlags: flags.AllowCall, ParamCount: 3},
-	{Name: interopnames.SystemContractCallEx, Func: contract.CallEx, Price: 1 << 15,
-		RequiredFlags: flags.AllowCall, ParamCount: 4},
+		RequiredFlags: flags.AllowCall, ParamCount: 5},
```
Comment (roman-khimov) [inline pkg/core/interops.go:54]: Isn't it 4 now (hash, method, flags, args)?

File: pkg/vm/emit/emit.go [hunk 3]
```diff
@@ -156,13 +157,19 @@ func AppCall(w *io.BinWriter, scriptHash util.Uint160) {
 	Syscall(w, interopnames.SystemContractCall)
 }
 
-// AppCallWithOperationAndArgs emits an APPCALL with the given operation and arguments.
-func AppCallWithOperationAndArgs(w *io.BinWriter, scriptHash util.Uint160, operation string, args ...interface{}) {
+// AppCallWithFlag emits an APPCALL with the full set of parameters.
+func AppCallWithFlag(w *io.BinWriter, scriptHash util.Uint160, operation string, f flags.CallFlag, args ...interface{}) {
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:161]: Do we really need `emit.AppCall`? I think this can be a substitute for it.

File: pkg/vm/emit/emit.go [hunk 4]
```diff
@@ -156,13 +157,19 @@ func AppCall(w *io.BinWriter, scriptHash util.Uint160) {
...
 	Array(w, args...)
+	Int(w, int64(f))
 	String(w, operation)
 	AppCall(w, scriptHash)
 }
 
+// AppCallWithOperationAndArgs emits an APPCALL with the default parameters given operation and arguments.
+func AppCallWithOperationAndArgs(w *io.BinWriter, scriptHash util.Uint160, operation string, args ...interface{}) {
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:169]: `AppCallAllFlag`? Maybe it's easier to have one `AppCall`?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: cli/smartcontract/smart_contract.go [hunk 5]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/io"
 	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result"
 	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+	flags2 "github.com/nspcc-dev/neo-go/pkg/smartcontract/flags"
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go:27]: Fixed.

File: pkg/smartcontract/flags/call_flags_test.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package smartcontract
+package flags
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/smartcontract/flags/call_flags_test.go:1]: Fixed.

File: cli/testdata/deploy/main.go [hunk 7]
```diff
@@ -33,10 +33,11 @@ func Fail() {
 }
 
 // Update updates contract with the new one.
-func Update(script, manifest []byte) {
+func Update(script, manifest []byte) bool {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cli/testdata/deploy/main.go:36]: No, it is an artifact of older iteration, fixed.

File: pkg/compiler/syscall.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ var syscalls = map[string]map[string]string{
 	},
 	"contract": {
 		"Call":                  interopnames.SystemContractCall,
-		"CallEx":                interopnames.SystemContractCallEx,
+		"CallNoReturn":          interopnames.SystemContractCall,
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/compiler/syscall.go:26]: Fixed.

File: pkg/core/interops.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/interops.go:54]: Fixed.

File: pkg/vm/emit/emit.go (hunk 3 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go]: Fixed.

File: pkg/vm/emit/emit.go (hunk 4 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go]: Replaced this with `AppCall` and `AppCallNoArgs`. Latter is used in multiple tests and `rpc`, when arguments are already on stack.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**